### ISSUE TITLE
Misleading Behavior of model.evaluate()

### ISSUE THREAD

**Issue by profPlum** [keras-team/keras#16461]
I experienced a situation where model.evaluate() _was very misleading on model performance due to the default of using a batch_size=32_. I was using R^2 to evaluate my model and was getting quite low R^2 values from model.evaluate() compared to manual evaluation. **While on average the values were 5% lower than the true values, due to a few _extreme outlier mini-batches_ it once gave R^2~=-90,000 (for an entire dataset) when the true R^2~=95%!!!**

```
# R^2 metric code for reference
def R2(yt,yp): return tf.reduce_mean(1-tf.reduce_mean((yp-yt)**2, axis=0)/tf.math.reduce_variance(yt,axis=0))
```

I eventually realized it was because the mini-batches were causing the variance, to always be **sample variance** and that then the metric was being averaged after all the mini-batch computations. This is not valid for R^2! _It requires the best estimate of variance that is available otherwise it can create extremely erroneous results!_

Please, Please reconsider having the default option being set to use mini-batch evaluation. Not only is the default batch_size=None (which seems to imply no mini-batches would be used), it also only returns 1 value which makes you think only 1 evaluation of the metric took place.

P.S. I realize mini-batch evaluation is a documented default. I'm asking that you change the default behavior so it has less surprising negative side-effects (which can be very difficult to debug).

**Comment by profPlum** [comment]
P.P.S. If the changes I'm suggesting are too radical please at least consider this: using a larger default batch size for evaluation (e.g. 128) & _explicitly advertise the batch size_ by having the default batch_size!=None.

**Comment by tilakrayal** [comment]
Hello @profPlum ,
In order to expedite the trouble-shooting process, could you please provide a complete code and the TensorFlow version you are using.Thanks!

**Comment by profPlum** [comment]
@tilakrayal Hello, I'm not sure that is practical. The code base is very large as is the dataset. But I have confirmed that my diagnosis of the problem is correct (I was able to reproduce my exact manual evaluation of R^2 using model.evaluate() by setting the batch_size=n_dataset_rows). So I'm sure the reason for the discrepancy is the batch_size. 

The rest of the problem is a question of design of keras. And not related to my code.

**Comment by tilakrayal** [comment]
@profPlum ,
Without the reproducible code, it would be difficult for us to debug the issue. In order to expedite the trouble-shooting process, could you please provide a minimal code snippet or colab gist and the TensorFlow version you are using.Thanks!

**Comment by profPlum** [comment]
> @profPlum ,
> Without the reproducible code, it would be difficult for us to debug the issue. In order to expedite the trouble-shooting process, could you please provide a minimal code snippet or colab gist and the TensorFlow version you are using.Thanks!

There is no debugging required… I’ve already confirmed that Keras is working as documentation says it should….

The problem is the way that it works can be extremely misleading (please reread my previous reply).

**Comment by qlzh727** [comment]
From the description of the issue, I don't think it is the issue for default batch_size (which we clearly document the default value and behavior), I think it is more because the misunderstanding of the metric API.

keras metric is always an stateful object, which means it tracks the state across the batches, and aggregate the value.  If your metric need to aggregate differently, then you should code in that way (eg you should have count for how many examples you have seen, and reduced value to aggregate). 

If you want your eval result to be stateless, you could make it as a loss function (which is always stateless and apply to mini batch). You can use that for the eval only.

**Comment by profPlum** [comment]
@qlzh727 
1. Like I said I think it is very (& unnecessarily) counter intuitive, if you wanted it to be "very clear" I'd encourage you to set the default batch size not to None but the _actual default batch size_ (i.e. 32). So that it is "very clear" you are actually using mini-batches by default.
2. I still don't understand why mini batches are necessary. But if you insist on using them for evaluation as well: I'd also strongly encourage you to at least increase the default batch size to be 256 for example. R^2 is probably the most information metric for regression problems, and as I mentioned **previously just 1 mini-batch with sample variance 0 is enough to have evaluate() return R^2=-infinity for the entire dataset**.

I'd appreciate you not writing this off as me just 'not understanding the api' (as I mentioned already, I know how it works), I'm highlighting extremely surprising and misleading behavior that is actually quite likely to happen often (R^2 is the best regression metric IMO), in a very common use case of keras (simple feedforward regression is probably the most basic application).

**Comment by qlzh727** [comment]
@profPlum. 

1. The Keras fit/eval/predict has been using default batch size since it started, and we can't change the default value easily for backward compat reason. Note that a metric class should work regardless the batch size changes (otherwise it doesn't meet the API contract for metric). If keras.metric.Accuracy need to only work with a certain batch size, then it will be very confusing to end user.

2. Keras use mini batch for fit/eval/predict since the beginning of the project. Keras doesn't necessary know the whole data input size, and using whole data input to do fit/eval will quickly lead to out-of-memory error.

3. If you could provide your implementation of R2 (which u feed to model.eval as a metric), we probably can have further discussion. But from what you have described so far, the implementation doesn't aggregate the value correctly.
